content/2017.02.handling-python-projects-dependencies.rst

```diff
@@ -0,0 +1,225 @@
+
+This brinks a lot of questions:
+
+ - How do I update project dependencies?
+ - What were the know working dependencies version at a certain release?
+ - How do I make sure a release of one of the dependency I use doesn't break my release?
+ - How do I make sure that my code is not used with a non supported version of a dependency?
+
+
+Avoid using requirements file to handle your program dependencies
+=================================================================
+
+Because your program will eventually be released, you want to be able
+to use ``python setup.py install`` to install it so that you will be
+able to install it with ``pip install saucisson`` to install the last
+release and it will install itself automatically with the right
+dependencies.
+
+Instead use your ``requirements.txt`` to set a know working set of your
+dependencies at the time of a release.
+
+We usually add a `build-requirements` makefile rule on our projects:
+
+.. code-block:: Makefile
+
+    VIRTUALENV = virtualenv --python=python3
+    
+    
+    build-requirements:
+    	$(eval TEMPDIR := $(shell mktemp -d))
+    	$(VIRTUALENV) $(TEMPDIR)
+    	$(TEMPDIR)/bin/pip install -U pip  # Upgrade pip
+    	$(TEMPDIR)/bin/pip install -Ue .   # Develop the current project locally
+    	# Freeze the dependencies ignoring the dependency links.
+    	$(TEMPDIR)/bin/pip freeze | grep -v -- '^-e' > requirements.txt
+
+If you run `make build-requirements` before tagging your release you
+will document what was the known working dependency set at the time of
+the release.
+
+Then you can install your program later using this file as a
+`dependency constrain file <https://pip.pypa.io/en/stable/user_guide/#constraints-files>`_:
+
+.. code-block:: console
+
+    pip install saucisson -c requirements.txt
+
+
+Avoid pining version in your setup.py
+=====================================
+
+You might be tempted to pin dependencies version in your setup.py
+
+Because you are using a contrain file you are safe for future updates
+that might break your code.
+
+On your CI not using a contrain file will help you to detect that you
+need to take some actions to support the new released version.
+
+There are some cases when you still want to pin some version:
+
+
+If you know that your project will not work with a lower version
+----------------------------------------------------------------
+
+If you are using a feature or API that didn't existed before or you
+had a bug that was fixed after.
+
+.. code-block::
+
+   psycopg > 2.5
+   colander >= 1.3.2
+   ujson >= 1.35
+
+
+If you know that your project doesn't support yet the next release
+------------------------------------------------------------------
+
+**I insist that this implies that the release already exists and that
+your test suite doesn't run on it.**
+
+In that case, and only in that case, you can do it for the shorted
+possible time until you ported your project to it.
+
+It is often the case for frameworks:
```

Comment:
`It` is pretty vague here, and `often the case` doesn't sound right either. At the very least you could say `This situation happens often with frameworks`, but I think you could go further and say something like `Frameworks often need to be pinned for this reason` or `It's common to encounter breaking changes when upgrading frameworks`.